ambry-account/src/main/java/com/github/ambry/account/RouterStore.java

```diff
@@ -277,15 +284,33 @@ public ZNRecord update(ZNRecord znRecord) {
         accountServiceMetrics.accountUpdateToAmbryTimeInMs.update(System.currentTimeMillis() - startTimeMs);
       } catch (Exception e) {
         accountServiceMetrics.accountUpdatesToAmbryServerErrorCount.inc();
-        errorMessage =
-            "Updating accounts failed because unexpected error occurred when uploading AccountMetadata to ambry";
-        logger.error(errorMessage, e);
-        throw new IllegalStateException(errorMessage, e);
+        logAndThrowIllegalStateException(
+            "Updating accounts failed because unexpected error occurred when uploading AccountMetadata to ambry", e);
       }
 
       // Start step 5:
-      accountBlobIDs.add(new BlobIDAndVersion(this.newBlobID, newVersion).toJson());
-      recordToUpdate.setListField(ACCOUNT_METADATA_BLOB_IDS_LIST_KEY, accountBlobIDs);
+      if (blobIDAndVersions.size() + 1 > totalNumberOfVersionToKeep) {
+        Iterator<BlobIDAndVersion> iter = blobIDAndVersions.iterator();
+        while (blobIDAndVersions.size() + 1 > totalNumberOfVersionToKeep) {
+          BlobIDAndVersion blobIDAndVersion = iter.next();
+          iter.remove();
+          logger.info("Removing blob " + blobIDAndVersion.getBlobID() + " at version " + blobIDAndVersion.getVersion());
+          try {
+            router.get().deleteBlob(blobIDAndVersion.getBlobID(), SERVICE_ID).get();
```

Comment:
Are there any edge cases where we would not want to delete the oldest blob from within the update logic. 
The main case to consider is probably where the update to zookeeper does not succeed. In this case we would have deleted a blob in ambry that could still exist in the current znode.

However, this may not be a problem if `totalNumberOfVersionsToKeep` is sufficiently large that the chance of someone else still interested in reading that blob is very small.

Reply:
delete the blobs after successfully updates zookeeper node.